### ISSUE TITLE
Add a flag to turn on/off irrelevance (and disable it by default?)

### ISSUE THREAD

**Issue by jespercockx** [agda/agda#6246]
In https://github.com/agda/agda/issues/6049#issuecomment-1293718186, @martinescardo noted that he would like to have a flag to turn off irrelevance, and I agree that it would be nice to have such a flag (and its negation `--no-irrelevance`). I'm less sure about the question of whether this flag should be on or off by default. On one hand, irrelevance only kicks in when the user uses it explicitly somewhere. On the other hand, the syntax is very minimal (just a dot) so a new user might use it by accident. So having it turned off by default might make sense (but would break a lot of backwards compatibility).

**Comment by nad** [comment]
I guess the flag should be infective.

Cubical Agda uses irrelevance in some way, right? Would `--cubical-compatible` have to imply this flag? In that case the flag might not be very useful, because several built-in modules use `--cubical-compatible`. Perhaps one could turn on the flag locally, in a non-infective way, when the generated Cubical Agda clauses are checked.

The standard library also uses irrelevance, for instance to define empty types, so most code that uses the standard library would probably have to turn on this flag.

**Comment by andreasabel** [comment]
Nothing speaks against flags to turn off language features of Agda.
But irrelevance seems to be a stable feature (with problematic extensions hidden under flags that are off by default), so I do not see why to turn it off by default.  Purists can turn it off if they like.

**Comment by martinescardo** [comment]
It is not a matter of purity. If you say you proved something in a mathematical or logical journal,  and if you are working in a non-standard foundation, the referees will want to know precisely what foundation you are using. And features that "seem to be stable" are probably not going to be very convincing.

**Comment by andreasabel** [comment]
Agda has many uses.  Breathing life into MLTT is one (but we don't have a `--mltt` flag yet).  Others might just care for a dependently typed programming language.  There is no clear _default_ use of Agda.

**Comment by martinescardo** [comment]
I'd love an `--mltt` flag.

**Comment by jespercockx** [comment]
What would an `--mltt` flag even mean? No pattern matching but only eliminators? No data types other than empty, unit, bool, nat, sigma, sum, and W? No mutual definitions? No universe polymorphism? No copatterns? I think it's better to have flags that have a clear semantics of what exactly they turn on or off.

EDIT: An alternative interpretation would be that `--mltt` should turn off all features that could not be *encoded* in MLTT, so we could still have some restricted forms of data types and pattern matching. I still think that mutual definitions should probably be out (which would require a new semantics for `where` blocks as well). And perhaps a different more restricted form of universe polymorphism could be allowed.

**Comment by martinescardo** [comment]
I favour the alternative interpretation. But of course this would need a more careful discussion.

**Comment by martinescardo** [comment]
Also, as I said elsewhere, given that Agda is addressed to a wide audience (programmers, computer scientists, logicians, mathematicians), then this is precisely why we should be able to tune it to the specific wishes of the niches among the wide audiences. There is no default that will accommodate the wishes of such a general audience. At the moment the defaults seem a little bit random.

**Comment by martinescardo** [comment]
Oh, and I forgot to mention beginners in my list. For them, the fewer the enabled features, the better. I know because I teach Agda (at MGS, at HoTTEST, and at my own institution as an undergrad course).

**Comment by plt-amy** [comment]
> Cubical Agda uses irrelevance in some way, right?

Yes, but not in an essential way. We use it in the definition of `Partial`/`PartialP`: `Partial φ A` is defined to be almost (but [not quite](https://github.com/agda/agda/issues/6015)) like `.(IsOne φ) → A`, as an alternative to putting `IsOne` in `Prop`.

**Comment by andreasabel** [comment]
Agda dev meeting 2024-05-02: we could have the `--irrelevance` flag, but maybe keep it on by default for now.
